std/assembly/util/uri.ts

```diff
@@ -0,0 +1,266 @@
+          offset += 6;
+          storeHex(dst, offset, (c >> 12 & 0x3F) | 0x80);
+          offset += 6;
+        }
+        storeHex(dst, offset, (c >> 6 & 0x3F) | 0x80);
+        offset += 6;
+      }
+      storeHex(dst, offset, (c & 0x3F) | 0x80);
+      offset += 6;
+    }
+    ++i;
+  }
+  if (outSize > offset) {
+    dst = __renew(dst, offset);
+  }
+  return dst;
+}
+
+export function decode(src: usize, len: usize, component: bool): usize {
+  if (!len) return src;
+
+  var i: usize = 0, offset: usize = 0, org: usize, ch: u32 = 0;
+  var dst = __new(len << 1, idof<String>());
+
+  while (i < len) {
+    org = i;
+    while (i < len && (ch = load<u16>(src + (i << 1))) != CharCode.PERCENT) i++;
+
+    if (i > org) {
+      let size = i - org << 1;
+      if (size == 2) {
+        store<u16>(dst + offset, load<u16>(src + (org << 1)));
+      } else if (size == 4) {
+        store<u32>(dst + offset, load<u32>(src + (org << 1)));
+      } else {
+        memory.copy(
+          dst + offset,
+          src + (org << 1),
+          size
+        );
+      }
+      offset += size;
+      if (i >= len) break;
+    }
+
+    // decode hex
+    if (
+      i + 2 >= len ||
+      ch != CharCode.PERCENT ||
+      (ch = loadHex(src, i + 1 << 1)) == -1
+    ) throw new URIError(E_URI_MALFORMED);
+
+    i += 3;
+    if (ch < 0x80) {
+      if (!component && isReserved(ch)) {
+        ch = CharCode.PERCENT;
+        i -= 2;
+      }
+    } else {
+      // decode UTF-8 sequence
+      let nb = utf8LenFromUpperByte(ch);
+      // minimal surrogate: 2 => 0x80, 3 => 0x800, 4 => 0x10000, _ => -1
+      let lo: u32 = 1 << (17 * nb >> 2) - 1;
+      // mask: 2 => 31, 3 => 15, 4 => 7, _ =>  0
+      ch &= nb ? (0x80 >> nb) - 1 : 0;
+
+      while (--nb != 0) {
+        let c1: u32;
+        // decode hex
+        if (
+          i + 2 >= len ||
+          load<u16>(src + (i << 1)) != CharCode.PERCENT ||
+          (c1 = loadHex(src, i + 1 << 1)) == -1
+        ) throw new URIError(E_URI_MALFORMED);
+
+        i += 3;
+        if ((c1 & 0xC0) != 0x80) {
+          ch = 0;
+          break;
+        }
+        ch = (ch << 6) | (c1 & 0x3F);
+      }
+
+      if (ch < lo || lo == -1 || ch > 0x10FFFF || (ch >= 0xD800 && ch < 0xE000)) {
+        throw new URIError(E_URI_MALFORMED);
+      }
+
+      if (ch >= 0x10000) {
+        ch -= 0x10000;
+        let lo = ch >> 10 | 0xD800;
+        let hi = (ch & 0x03FF) | 0xDC00;
+        store<u32>(dst + offset, lo | (hi << 16));
+        offset += 4;
+        continue;
+      }
+    }
+    store<u16>(dst + offset, ch);
+    offset += 2;
+  }
+
+  assert(offset <= (len << 1));
+  if ((len << 1) > offset) {
+    dst = __renew(dst, offset);
+  }
+  return dst;
+}
+
+function storeHex(dst: usize, offset: usize, ch: u32): void {
+  // @ts-ignore: decorator
+  const HEX_CHARS = memory.data<u8>([
+    0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+    0x38, 0x39, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46
+  ]);
+
+  let hex =
```

Comment:
Only has one use and doesn't seem to contribute much to clarity much, so perhaps can be moved?